storage/sqlite/store.go

```diff
@@ -0,0 +1,204 @@
+/*
+ * Flow Emulator
+ *
+ * Copyright 2019 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sqlite
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	_ "github.com/glebarez/go-sqlite"
+	"github.com/onflow/flow-emulator/storage"
+)
+
+var _ storage.SnapshotProvider = &Store{}
+
+// Store implements the Store interface
+type Store struct {
+	storage.DefaultStore
+	db  *sql.DB
+	url string
+	mu  sync.RWMutex
+}
+
+func (s *Store) ListSnapshots() (snapshots []string, err error) {
+	if !s.SupportSnapshotsWithCurrentConfig() {
+		return []string{}, fmt.Errorf("Snapshot is not supported with memory db")
+	}
+
+	path := filepath.Dir(s.url)
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return snapshots, err
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		if strings.HasPrefix(file.Name(), "snapshot_") {
+			snapshotName := strings.TrimPrefix(file.Name(), "snapshot_")
+			snapshots = append(snapshots, snapshotName)
+		}
+
+	}
+	return snapshots, nil
+}
+
+func (s *Store) JumpToSnapshot(snapshotName string, createIfNotExists bool) error {
+	if !s.SupportSnapshotsWithCurrentConfig() {
+		return fmt.Errorf("Snapshot is not supported with memory db")
+	}
+	path := filepath.Dir(s.url)
+	dbfile := fmt.Sprintf("%s/snapshot_%s", path, snapshotName)
+
+	_, err := os.Stat(dbfile)
+	if !createIfNotExists && os.IsNotExist(err) {
+		return fmt.Errorf("Snapshot %s does not exist", snapshotName)
+	}
+
+	if os.IsNotExist(err) {
+		_, err := s.db.Exec(fmt.Sprintf("VACUUM main INTO '%s'", dbfile))
+		if err != nil {
+			return err
+		}
+	}
+
+	db, err := sql.Open("sqlite", dbfile)
+	if err != nil {
+		return err
+	}
+
+	err = initdb(db)
+	if err != nil {
+		return err
+	}
+
+	//switch
+	s.db.Close()
+	s.db = db
+
+	return nil
+}
+
+func (s *Store) SupportSnapshotsWithCurrentConfig() bool {
```

Comment:
nit:
```suggestion
func (s *Store) SupportsSnapshots() bool {
```

Reply:
some backends support snapshot but not with the config ( for example when we use --sqlite-db deniz.sqlite or memory fix when using memory ) but I can rename

Reply:
Yeah I get that, but does it matter to the consumer why?

Reply:
I think it does not matter in general, normally storage implementing "SnapshotProvider" signals it supports snapshots, but there are some cases it does not due to config. So being `SnapshotProvider` not supporting `snapshots` case seemed confusing to me, so I decided to explain why in API. But can be shortened.